Fix events propagation to Home Assistant

mkowalczuk opened the pull request:
Fix events propagation to Home Assistant
    
`payload_sequence` (an advanced boolean option) controls whether zigbee2mqtt
should add an additional `sequence` integer attribute to payload. It starts
with 1 and increases for identical payloads allowing Home Assistant to
distinguish consecutive same events. Otherwise, all events with the same
payload (except the 1st) are ignored and don't trigger state change. Because of
that a workaround (use of mqtt trigger in automations) is required, e.g.
```
      trigger:
        platform: mqtt
        topic: 'zigbee2mqtt/bedroom_switch'
      condition:
        condition: template
        value_template: {{ 'single' == trigger.payload_json.click }}
```
After this change (and setting `payload_sequence = true` in the configuration
file, advanced section), the above example trigger could be converted to:
```
      trigger:
        platform: state
        entity_id: sensor.bedroom_switch
      condition:
      - condition: template
        value_template: >
          {{ trigger.to_state.state == 'single' and
             trigger.from_state.state != 'unavailable' }}
```
(Checking from_state is needed to ignore state change after Home Assistant or
zigbee2mqtt restarts.)
    
Additionally, it should be possible to react only on selected events in
sequence (e.g. every third press of the button).

mkowalczuk committed:
Fix events propagation to Home Assistant

`payload_sequence` (an advanced boolean option) controls whether zigbee2mqtt
should add an additional `sequence` integer attribute to payload. It starts
with 1 and increases for identical payloads allowing Home Assistant to
distinguish consecutive same events. Otherwise, all events with the same
payload (except the 1st) are ignored and don't trigger state change. Because of
that a workaround (use of mqtt trigger in automations) is required, e.g.

  trigger:
    platform: mqtt
    topic: 'zigbee2mqtt/bedroom_switch'
  condition:
    condition: template
    value_template: {{ 'single' == trigger.payload_json.click }}

After this change (and setting `payload_sequence = true` in the configuration
file, advanced section), the above example trigger could be converted to:

  trigger:
    platform: state
    entity_id: sensor.bedroom_switch
  condition:
  - condition: template
    value_template: >
      {{ trigger.to_state.state == 'single' and
         trigger.from_state.state != 'unavailable' }}

(Checking from_state is needed to ignore state change after Home Assistant or
zigbee2mqtt restarts.)

Additionally, it should be possible to react only on selected events in
sequence (e.g. every third press of the button).

Koenkk commented:
This is great! 

For some reason it's not working yet for me. 

I've configured the following automation (running home assistant 0.75.1):

```yaml
- id: blaaa
  trigger:
    platform: state
    entity_id: sensor.0x00158d0001d8e1d2
    to: 'single'
  action:
    service: switch.toggle
    entity_id: switch.0x84182600000e9afd
```

But the automation is only triggered once the value changes (e.g. `single` to ` double`). The sequence number is increased: https://imgur.com/a/pcnjIfO

What would be the problem here?

mkowalczuk commented:
Sorry for my late response, I was on vacation.

I changed the commit message – previous one contained a wrong example on how to use trigger, (defining desired state). To trigger automation on the same event, state should be omitted from the trigger and moved to the condition. The resulting automation code is actually longer than the workaround using MQTT platform, but I believe also more flexible.

Koenkk commented:
Thanks!

I just checked #218, and by only setting `force_update: true` I managed to get updates with the following automation:

```yaml
- id: blaaa
  trigger:
    platform: state
    entity_id: sensor.0x00158d0001d8e1d2
  condition:
    - condition: template
      value_template: {{ trigger.to_state.state == 'single' }}
  action:
    service: switch.toggle
    entity_id: switch.0x84182600000e9afd
```

The `force_update: true` seems to be an easier solution, have you also tried this?

mkowalczuk commented:
I just tried it and it works! It is indeed cleaner. :)

Koenkk commented:
Ok, thanks for verifying I will close this one and merge #218
